{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'bart2'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-2-9001a0b31386>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[1;32mimport\u001b[0m \u001b[0mbart2\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      2\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mos\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0msys\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[0mtypes\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mSimpleNamespace\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[1;31m# import from package\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mModuleNotFoundError\u001b[0m: No module named 'bart2'"
     ]
    }
   ],
   "source": [
    "import bart2\n",
    "import os,sys\n",
    "from types import SimpleNamespace\n",
    "\n",
    "# import from package\n",
    "from bart2 import OptValidator, ReadCount, RPRegress, EnhancerIdentifier, AUCcalc, StatTest, score_on_UDHS\n",
    "\n",
    "script_dir = os.path.dirname(os.path.realpath(__file__))\n",
    "ADAPTIVE_LASSO_MAXSAMPLES = 20 # TODO: should we fix it?\n",
    "\n",
    "def bart(options):\n",
    "    args = OptValidator.opt_validate(options)\n",
    " \n",
    "    # create output directory\n",
    "    try:\n",
    "        os.makedirs(args.outdir, exist_ok=True)\n",
    "    except:\n",
    "        sys.stderr.write('Output directory: {} could not be created. \\n'.format(args.outdir))\n",
    "        sys.exit(1)\n",
    "    sys.stdout.write(\"Output directory will be {} \\n\".format(args.outdir))\n",
    "    sys.stdout.write(\"Output file prefix will be {} \\n\".format(args.ofilename))\n",
    "\n",
    "    if args.species == 'hg38':\n",
    "        sys.stdout.write(\"Start prediction on hg38...\\n\")\n",
    "    elif args.species == 'mm10':\n",
    "        sys.stdout.write(\"Start prediction on mm10...\\n\")\n",
    "    sys.stdout.flush()\n",
    "\n",
    "    # bart geneset [-h] <-i genelist.txt> [--refseq] <-s species> [-t target] [-p processes] [--outdir] [options]\n",
    "    if args.subcommand_name == 'geneset':\n",
    "        '''\n",
    "        Use adaptive lasso regression to select H3K27ac samples.\n",
    "\n",
    "        RPRegress parameters:\n",
    "        species, rp matrix, gene file, refseq TSS, output directory, target gene method, adaptive lasso max sapmle numbers, log transform, square transform, gene symbol or refseqID, gene symbol or not, separate by chrom, sample description file\n",
    "        '''\n",
    "        sys.stdout.write(\"Do adaptive lasso to select informative H3K27ac samples...\\n\")\n",
    "        sys.stdout.flush()\n",
    "\n",
    "        sys.stdout.write(\"Generate parameters for regression step...\\n\")\n",
    "        if options.refseq:\n",
    "            rp_args = \\\n",
    "                SimpleNamespace(genome=args.species, \\\n",
    "                          histRP=args.rp, \\\n",
    "                          expr=args.infile, \\\n",
    "                          sym=args.tss, \\\n",
    "                          name=args.ofilename, \\\n",
    "                          maxsamples=ADAPTIVE_LASSO_MAXSAMPLES, \\\n",
    "                          transform='sqrt', \\\n",
    "                          exptype=\"Gene_Response\", \\\n",
    "                          annotation=args.desc)\n",
    "            RPRegress.main(rp_args)\n",
    "        else:\n",
    "            rp_args = \\\n",
    "                SimpleNamespace(genome=args.species, \\\n",
    "                          histRP=args.rp, \\\n",
    "                          expr=args.infile, \\\n",
    "                          sym=args.tss, \\\n",
    "                          name=args.ofilename, \\\n",
    "                          maxsamples=ADAPTIVE_LASSO_MAXSAMPLES, \\\n",
    "                          transform='sqrt', \\\n",
    "                          exptype=\"Gene_Only\", \\\n",
    "                          annotation=args.desc)\n",
    "            RPRegress.main(rp_args)\n",
    "\n",
    "        regression_info = args.ofilename + '_adaptive_lasso_Info.txt'\n",
    "        if not os.path.exists(regression_info):\n",
    "            sys.stderr.write(\"Error: selecting samples from H3K27ac compendium! \\n\")\n",
    "            sys.exit(1)\n",
    "\n",
    "        '''\n",
    "        Generate cis-regulatory profile based on adaptive lasso model weights multiply H3K27ac samples RPKM signal.\n",
    "\n",
    "        EnhancerIdentifier parameters:\n",
    "        selected samples file, gene file, output directory, species, UDHS, rpkm matrix\n",
    "        '''\n",
    "        sys.stdout.write(\"Generate cis-regulatory profile...\\n\")\n",
    "        sys.stdout.flush()\n",
    "\n",
    "        sys.stdout.write(\"Generate parameters for enhancer profile generation...\\n\")\n",
    "        enhancer_args = \\\n",
    "                SimpleNamespace(samplefile=regression_info, \\\n",
    "                                name=args.ofilename, \\\n",
    "                                k27achdf5=args.rpkm)\n",
    "        EnhancerIdentifier.main(enhancer_args)\n",
    "        sys.stdout.flush()\n",
    "\n",
    "        enhancer_profile = args.ofilename + '_enhancer_prediction_lasso.txt'\n",
    "        if not os.path.exists(enhancer_profile):\n",
    "            sys.stderr.write(\"Error: generating enhancer profile! \\n\")\n",
    "            sys.exit(1)\n",
    "\n",
    "        # get ranked score UDHS positions from enhancer profile\n",
    "        positions = AUCcalc.get_position_list(enhancer_profile)\n",
    "\n",
    "    # bart profile [-h] <-i ChIP-seq profile> <-f format> <-s species> [-t target] [-p processes] [--outdir] [options]\n",
    "    elif args.subcommand_name == 'profile':\n",
    "        sys.stdout.write('Start mapping the {} file...\\n'.format(args.format.upper()))\n",
    "        sys.stdout.flush()\n",
    "        counting = ReadCount.read_count_on_DHS(args)\n",
    "        # get ranked score UDHS positions from read count\n",
    "        positions = sorted(counting.keys(),key=counting.get,reverse=True)\n",
    "\n",
    "    elif args.subcommand_name == 'region':\n",
    "        sys.stdout.write('Start mapping the bed score onto UDHS...\\n')\n",
    "        sys.stdout.flush()\n",
    "        counting = score_on_UDHS.score_on_DHS(args)\n",
    "        sys.stdout.write('Sorting scored UDHS...\\n')\n",
    "        sys.stdout.flush()\n",
    "        positions = sorted(counting.keys(),key=counting.get,reverse=True)\n",
    "\n",
    "\n",
    "    '''\n",
    "    Start using revised BART on calculating the AUC score for each TF ChIP-seq dataset\n",
    "    '''\n",
    "    sys.stdout.write('BART Prediction starts...\\n\\nRank all DHS...\\n')\n",
    "    sys.stdout.flush()\n",
    "\n",
    "    tf_aucs, tf_index = AUCcalc.cal_auc(args, positions)\n",
    "    sys.stdout.flush()\n",
    "\n",
    "    stat_file = args.ofilename + '_bart_results.txt'\n",
    "    StatTest.stat_test(tf_aucs, tf_index, stat_file, args.normfile)\n",
    "    sys.stdout.flush()\n",
    "    sys.stdout.write(\"Congratulations! BART job finished successfully!\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
